### PR TITLE
[WIP] Support hiding components in devtools

### PR DESCRIPTION
**Pull request by Gvozd**
Add two options for hiding components
1) For components with `[Symbol.for('react.devtools.hide')] = true` - rebase of PR #997 
2) If component define `displayName` - I think that most popular case for HOC

I'm waiting for feedbacks
It is time to move forward on this issue after long discussions.

### CODE REVIEW

**Comment by Gvozd** [comment]
Call in the discussion of interested persons
@gaearon @sebmarkbage @Kovensky @zackargyle @kyleshevlin @jquense @jaredly @shriah @bvaughn

**Comment by jaredly** [comment]
Looks good to me! I'm no longer a committer on this project, so I'll have to let someone else merge

File: backend/attachRendererFiber.js [hunk 1]
```diff
@@ -176,6 +178,8 @@ function attachRendererFiber(hook: Hook, rid: string, renderer: ReactRenderer):
         publicInstance = fiber.stateNode;
         props = fiber.memoizedProps;
         state = fiber.memoizedState;
+        hideSymbol = typeof Symbol === 'function' && fiber.type[Symbol.for('react.devtools.hide')];
+        hideDisplayNamed = fiber.type.hasOwnProperty('displayName');
```
Comment (bvaughn) [inline backend/attachRendererFiber.js]: Sorry, I'm jet lagged at the moment so maybe this is a stupid question, but why is the "displayName" check here? I don't get how it's related.

File: frontend/Breadcrumb.js [hunk 2]
```diff
@@ -163,7 +163,7 @@ function getBreadcrumbPath(store: Store): BreadcrumbPath {
 }
 
 module.exports = decorate({
-  listeners: () => ['breadcrumbHead', 'selected'],
+  listeners: () => ['breadcrumbHead', 'selected', 'hideSymbol', 'hideDisplayNamed'],
```
Comment (bvaughn) [inline frontend/Breadcrumb.js]: Why is `Breadcrumb` listening for `hideDisplayNamed` and `hideSymbol` when it doesn't use them?

File: frontend/PreferencesPanel.js [hunk 3]
```diff
@@ -118,6 +122,26 @@ class PreferencesPanel extends React.Component<Props, State> {
...
+            </label>
+          </div>
+          <div>
+            <label>
+              <input
+                type="checkbox"
+                checked={hideDisplayNamed}
+                onChange={this._changeHideDisplayNamed}
+              />
+              Hide components with custom <code>displayName</code> property
```
Comment (bvaughn) [inline frontend/PreferencesPanel.js]: I don't understand why you'd want this particular functionality.

File: backend/attachRendererFiber.js [hunk 4]
```diff
@@ -176,6 +178,8 @@ function attachRendererFiber(hook: Hook, rid: string, renderer: ReactRenderer):
         publicInstance = fiber.stateNode;
         props = fiber.memoizedProps;
         state = fiber.memoizedState;
+        hideSymbol = typeof Symbol === 'function' && fiber.type[Symbol.for('react.devtools.hide')];
```
Comment (bvaughn) [inline backend/attachRendererFiber.js]: nit: `react-devtools.hide`?

**Comment by bvaughn** [comment]
I'm not sure how I feel about the hidden symbol. I don't like the `hideDisplayNamed` logic on initial read though. It seems unintuitive to me. Maybe you can explain your thought process?

File: frontend/PreferencesPanel.js (hunk 3 above)
Comment (Gvozd) [reply frontend/PreferencesPanel.js]: Many HOC's creates wrapper components with custom displayName like `SomeWrapper(OriginalComponent)`
In contrast, regular components usually do not use dispalyName

I thought that such a simple rule is suitable for a draft, and better than nothing
This is the minimum that I would like to get in order to hide noisy HOC's
I can add/replace with other simple HOC recognition, if they work better.

Comment (asvetliakov) [reply frontend/PreferencesPanel.js]: > Many HOC's creates wrapper components with custom displayName
>In contrast, regular components usually do not use dispalyName

There are many that are kind of like "presentation HOC" and assign displayName through babel plugin, i.e styled-components, emotion, jss, etc:
```jsx
const MyButton = styled.button`
   color: tomato;
`
// MyButton will have displayName: "MyButton" crafted by babel plugin
```
Although they are technically HOC i'd not want to hide them.

Comment (bvaughn) [reply frontend/PreferencesPanel.js]: I think the `displayName` attribute is too likely to cause false positives, as @asvetliakov points out, due to popular plug-ins like [`babel-plugin-transform-react-display-name`](https://www.npmjs.com/package/babel-plugin-transform-react-display-name). It also feels too arbitrary to me.

**Comment by bvaughn** [comment]
I believe I would also be more of a fan of that approach, as it would be more consistent with filtering other categories of components (as described in #1076)

**Comment by FredyC** [comment]
This was so good looking until it wasn't. Would you please remove the second approach with `displayName`? It's too much opinionated and not that useful. Even though it's opt-in, it would be just confusing.

**Comment by Gvozd** [comment]
Sorry for the break
Replaced `'Hide components with custom 'displayName' property` to `Hide HOC's with parens in name`
Fixed some bug with breadcrumbs, and selection
Corrected UI
image

In next plans:
- additional dimming mode
- optimize performance

**Comment by FredyC** [comment]
@Gvozd thanks, much better.

I am wondering, won't this symbol approach be somewhat problematic in TypeScript or possibly even Flow? Generally attaching foreign properties to objects is rather painful with static typing. Any ideas how it could be solved except using `// @ts-ignore` above the assignment line?

**Comment by Kovensky** [comment]
I wasn't thinking of attaching it to arbitrary functions but rather to functions/classes you generate yourself as part of the HOC factory function. `static [Symbol.for('react-devtools.hide')] = true`.

**Comment by FredyC** [comment]
Well, for an example I would like to hide all `React.createContext` so I suppose I would have to do something like this? 

```ts
const context = React.createContext()
// @ts-ignore
context.Consumer[hideSymbol] = true
```

And also some stateless components are useless to be visible in the tree, so that's about attaching a symbol to existing function.

**Comment by Gvozd** [comment]
1) I fixed the performance of many nested components
2) For custom static symbol/ordinary property
If they are approved by maintainers - need fix ts/flow lib definition for react
In this case, setting this property should work as well as `displayName`
It is also necessary to use an ordinary property, in order to properly support the flow
Offer `hideInDevtools` as the name of a static property

**Comment by Gvozd** [comment]
> TLDR: There are no plans to remove classes from React.
The old HOCs will not disappear anywhere for a long time
Classes do not go anywhere, so the hooks will live for a long time in parallel with the existing APIs, classes, and therefore HOC/render props

It also takes time
- stabilize RFC
- while the most popular libraries will add hoke API
In addition, because hooks do not work with class components, and no one is going to refuse classes
- when (if) existing projects will be transferred to hooks, and exclusively functional components

This is the case if *all* developers accept hooks as more convenient, and decide to switch to them in full
Until then (an optimistic assessment for the next year or two), the HOC will be a frequent guest in many already existing projects

So I think that to hide the HOCis still necessary and useful.

**Comment by Gvozd** [comment]
Upd: support dimming
At next - I fix bug with higlighting elements in page, when Preferences panel is opened
image

**Comment by Gvozd** [comment]
I think I need to leave only one option of hiding/ darkening, for the HOC with parentheses in the name
It is also worth moving these options to SettingsPane, as in #503, only in the form of selectbox, so as not to take up much space
What you think?

**Comment by Inve1951** [comment]
>I think I need to leave only one option of hiding/dimming, for the HOC with parentheses in the name

I think dimming should be the default. A toggle to hide instead would be usefull.
So basically, off = dimmed, on = hidden.

>It is also worth moving these options to SettingsPane, as in #503, only in the form of selectbox, so as not to take up much space

Good idea.

**Comment by bvaughn** [comment]
I don't think the dimming is very useful. It's very subtle and the tree is just as cluttered as before. (I think the primary purpose of hiding components is that the element tree is too crowded.)

I also think it's confusing to have a checkbox that says, e.g. "Hide HOCs..." when they won't actually be hidden unless you select _one of three_ radio options above. I suppose I could imagine a UX where we have a global toggle on/off, but I don't think it's necessary.

I realize this PR has been pretty long running, and I feel bad about the inconvenience. If you'd like, I could pick up where this PR leaves off and make the additions I had planned for #1076 originally?

**Comment by Gvozd** [comment]
@bvaughn, OK.

**Comment by Inve1951** [comment]
>I don't think the dimming is very useful. It's very subtle and the tree is just as cluttered as before. (I think the primary purpose of hiding components is that the element tree is too crowded.)

I personally appreciate a dimming option for HOCs, as they are part of the virtual DOM and I'd like to view them as such, but at the same time want them to be visually distinguishable/faded since they are HOCs / less impactful than regular components.

>I also think it's confusing to have a checkbox that says, e.g. "Hide HOCs..." when they won't actually be hidden unless you select one of three radio options above. I suppose I could imagine a UX where we have a global toggle on/off, but I don't think it's necessary.

We just talked about this and I agree that the radio buttons are inconvenient. My suggestion was to make a checkbox/toggle for hiding. If dimming gets included in this feature then it could be the off-state of said toggle.

@bvaughn @Gvozd
